src/ocean/application/Daemon.d

```diff
@@ -0,0 +1,814 @@
+    import ocean.application.components.Timers;
+    import ocean.application.components.UnixSocketCommands;
+    import Version = ocean.application.components.Version;
+
+    /// Settings passed to the constructor.
+    public struct Settings
+    {
+        /// Application name. Mandatory.
+        istring name;
+
+        /// Application description string.
+        istring desc;
+
+        /// How the program is supposed to be invoked.
+        istring usage;
+
+        /// Long description of what the program does and how to use it.
+        istring help;
+
+        /// Application version info. Mandatory.
+        Version.VersionInfo ver;
+
+        /// Configuration files to parse.
+        istring[] config_files = [ "etc/config.ini" ];
+
+        /// If true, errors when parsing configuration files are treated as
+        /// fatal -- the app will not start.
+        bool config_errors_fatal = true;
+
+        /// Set of signals to handle.
+        int[] signals;
+
+        /// Set of signals to ignore. Delivery of the signals specified in this
+        /// set will have no effect on the application -- they are not passed to
+        /// the default signal handler.
+        int[] ignore_signals;
+
+        /// Default scheduler configuration to be used if no scheduler settings
+        /// are specified in the config file.
+        Scheduler.Configuration scheduler_config;
+    }
+
+    /// Settings passed to the constructor.
+    protected Settings settings;
+
+    /// Command line args.
+    protected Arguments args;
+
+    /// Configuration parser to use to parse the configuration files.
+    protected ConfigParser config;
+
+    /// Stats logger.
+    protected StatsLog stats_log;
+
+    /// Stats logging config. (Also required for timer setup.)
+    private StatsLog.Config stats_log_config;
+
+    /// Set of files opened by the application that should be reopened on signal.
+    protected OpenFiles files;
+
+    /// Unix socket helper.
+    protected UnixSocketCommands unix_socket;
+
+    /// Application-level timers.
+    protected Timers timers;
+
+    /// PID lock helper.
+    private PidLock pid;
+
+    /// Signal handlers.
+    private Signals signals;
+
+    /// Process stats tracker.
+    private CpuMemoryStats process_stats;
+
+    // TODO: global process monitor (see ocean.io.select.client.EpollProcess)
+
+    /// Arrays of delegates that are called at certain points in the app
+    /// start-up procedure.
+    private struct Hooks
+    {
+        /// Hooks called just before args handling has occurred.
+        void delegate ( )[] pre_handle_args;
+
+        /// Hooks called during validation of args. If a validation error
+        /// occurs, a description of the error should be appended to the array
+        /// of strings passed to the delegate.
+        void delegate ( istring[] )[] validate_args;
+
+        /// Hooks called to process the parsed config files.
+        void delegate ( )[] process_config;
```

Comment:
No `pre_run` hooks?

Reply:
I didn't find a use for them, in my adaptation of the dhtnode at least. Anything "pre-run" can equally easily be done first thing inside `run()`. Can you think of something specific that couldn't work that way?

Reply:
I was thinking about it in context of https://github.com/sociomantic-tsunami/ocean/issues/500 - right now recommended way to initialize app is in `processConfig` override but it would be nice to support dedicated spot for that in a new system.